docs/design/storage-auto-scale.md

```diff
@@ -0,0 +1,94 @@
+
+- If the OSD size is equal to 8Tib, do a horizontal scaling, by adding 1 osd of 8Tib on each storage node, by simply increasing the (count+1) in the device set associated with a device class.   
+
+# Low-Level Design 
+
+1) Add a single go-routine to watch periodically(10mins) all the OSDs sizes using Prometheus metrics in the storagecluster namespace.
+   1) With every run calculate different device class highest osd filled percentage.
+       `OsdPercentage`:
+        ```
+        ssd               69%
+        nvme               71%
+        ...
+        ```
+    2) If the highest osd percentage reported is more than 73-74%(reaching osd nearfull), scaling is needed.
+    3) If scaling is needed calculate the `expectedOSDSize` and `expectedOSDCount` per device class.
+       1)  Calculate the total number osds and osd size per device class from the prometheus response.
+       2)  Calculate the total number osds and osd size per device class from the storageCluster.
+       3)  If the above values are equal(needed to avoid double re-sizing), create or update the sync-map mapping:
+           1)  If the OSD size is less than 8Tib, do vertical scaling by doubling the each osd sizes for that device class.
+           2)  If the OSD size is equal to 8Tib, do a horizontal scaling, by adding 1 osd of 8Tib on each storage node.
+  
+    4) Go routine would update the [sync-map](https://go.dev/src/sync/map.go) with different device class expected state.
+
+        `deviceClassScaling` sync-map:
+
+        ```
+        ssd:
+            expectedOSDSize: 4
+            expectedOSDCount: 6
+        hdd:
+            expectedOSDSize: 4
+            expectedOSDCount: 9
+        ...
+        ```
+
+2) If the sync map is created or updated then go routine will send a signal to the go channel.
+
+3) Create a named controller that watches for [channel generic event](https://book-v1.book.kubebuilder.io/beyond_basics/controller_watches).
+
+4) If there is an update on the channel, watcher will reconcile the new named controller.
+   1) Named Controller will list all the StorageAutoScaling CRs.
+   2) Depending on the `OsdPercentage sync-map` it will set the `expectedOSDSize` and `expectedOSDCount` and on the status and update `phase` to `in-progress` on the `StorageAutoScaling` CR which need scaling.
+   3) Check if the totalCapacity allows to scale it or not.
+   4) Patch the Storagecluster, with all the device class device sets update needed at the same time.
+   5) Verify the Storagecluster whether the new osds are added or scaled in size, for all the device class device sets. 
+   6) If the scaling is successful will raise a `success` alert and update the status of all the `StorageAutoScaling` CR.
+   7) If it is fails will do a re-reconcile and, if 3 recent reconcile fails(track by observed generation) will raise a `failure` alert.   
+ 
+5) Sync map will act as the desired state and source of truth for the controller.
```

Comment:
This statement is a little concerning. The desired state source of truth should be the CRD spec and status. The map is just an internal mechanism to keep track of info. I would suggest not keeping the map and instead rebuilding it with each reconcile (we will need to look at prometheus metrics each reconcile no matter what anyway). That will help prevent against bugs that result from caching stale data that never gets corrected.

Reply:
Yes we would be rebuilding the map every 10 min, 

But not on the cases where we updated the storagecluster and the storagcluster is in pending state to resize the storage

Reply:
@BlaineEXE There are 2 sources of truth here, each one is of its own domain (one is configuration, the other one is that state of the system) and we need to trigger the reconcile whenever either one changes.

Reply:
Once design is updated from the latest discussions, I don't think we need a syncmap, as no async communication is needed. Each 10 min, the reconcile just rebuilds the latest state.